### PR TITLE
feat: distribute truncate table in region server

### PR DESCRIPTION
**Pull request by DevilExileSu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
Implement truncate table in region server.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -78,7 +78,7 @@ datafusion-substrait = { git = "https://github.com/waynexia/arrow-datafusion.git
 derive_builder = "0.12"
 futures = "0.3"
 futures-util = "0.3"
-greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "115c1080773be8a819e50b257fece9f839a0c836" }
+greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "ee08f04ee2fcbc7dc4f722eab9e7eb1f120f4e1a" }
```
Comment (evenyag) [inline Cargo.toml:81]: The latest proto already has region truncate request.

File: src/common/meta/src/ddl/truncate_table.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+        for datanode in leaders {
+            let clients = self.context.datanode_manager.clone();
+            let regions = find_leader_regions(region_routes, &datanode);
+            let region_ids = regions
+                .iter()
+                .map(|region_number| RegionId::new(table_id, *region_number))
+                .collect::<Vec<_>>();
+
+            truncate_region_tasks.push(async move {
```
Comment (evenyag) [inline src/common/meta/src/ddl/truncate_table.rs:152]: Maybe we should create a task for each region.
https://github.com/GreptimeTeam/greptimedb/blob/00272d53cc0e5d7638e214a6160a1c727907fdcf/src/common/meta/src/ddl/alter_table.rs#L203-L223

File: src/common/meta/src/ddl/truncate_table.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                            trace_id: 0,
+                            span_id: 0,
+                        }),
+                        body: Some(region_request::Body::Truncate(PbTruncateRegionRequest {
+                            region_id: region_id.as_u64(),
+                        })),
+                    };
+
+                    if let Err(err) = clients.datanode(&datanode).await.handle(request).await {
+                        if err.status_code() != StatusCode::RegionNotFound {
```
Comment (evenyag) [inline src/common/meta/src/ddl/truncate_table.rs:167]: suggestion(non-blocking): Should we add a warning if status code is region not found?

File: src/common/meta/src/ddl/truncate_table.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            let clients = self.context.datanode_manager.clone();
+            let regions = find_leader_regions(region_routes, &datanode);
+            let region_ids = regions
+                .iter()
+                .map(|region_number| RegionId::new(table_id, *region_number))
+                .collect::<Vec<_>>();
+
+            truncate_region_tasks.push(async move {
+                for region_id in region_ids {
+                    debug!("Truncating region {region_id} on Datanode {datanode:?}");
```
Comment (evenyag) [inline src/common/meta/src/ddl/truncate_table.rs:154]: Also print the table name or table ref if we can get it.

File: tests/cases/standalone/common/truncate/truncate.sql [hunk 5]
```diff
@@ -0,0 +1,27 @@
+CREATE TABLE monitor (host STRING, ts TIMESTAMP, cpu DOUBLE DEFAULT 0, memory DOUBLE, TIME INDEX (ts), PRIMARY KEY(host));
```
Comment (evenyag) [inline tests/cases/standalone/common/truncate/truncate.sql:1]: suggestion: Also add a test to truncate a table that doesn't exist

File: src/common/meta/src/ddl/truncate_table.rs (hunk 3 above)
Comment (DevilExileSu) [reply src/common/meta/src/ddl/truncate_table.rs]: There might be a problem here, `RegionNotFound` should not be filtered out.

Comment (WenyXu) [inline src/common/meta/src/ddl/truncate_table.rs]: I don't think we will incur `RegionNotFound`. Any procedures executed in `metasrv`, acquire a table name lock.

File: src/common/meta/src/ddl/truncate_table.rs (hunk 2 above)
Comment (WenyXu) [inline src/common/meta/src/ddl/truncate_table.rs]: How about executing all region requests in parallel?

File: tests/cases/standalone/common/truncate/truncate.sql [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+(1695217654000, 'host1', 77.7, 2048),
+(1695217654000, 'host2', 77.7, 2048),
+(1695217654000, 'host3', 77.7, 2048),
+(1695217656000, 'host1', 88.8, 4096),
+(1695217656000, 'host2', 88.8, 4096),
+(1695217656000, 'host3', 88.8, 4096);
+
+SELECT ts, host, cpu, memory FROM monitor ORDER BY ts;
+
+TRUNCATE monitor;
```
Comment (WenyXu) [inline tests/cases/standalone/common/truncate/truncate.sql:18]: What if the user truncates the table many times? And should we add tests like:

```
TRUNCATE monitor;
INSERT INTO ...
TRUNCATE monitor;
```

File: src/common/meta/src/ddl/truncate_table.rs (hunk 3 above)
Comment (DevilExileSu) [reply src/common/meta/src/ddl/truncate_table.rs]: Yes, you are right.

File: tests/cases/standalone/common/truncate/truncate.sql (hunk 6 above)
Comment (DevilExileSu) [reply tests/cases/standalone/common/truncate/truncate.sql:18]: Thank you for your suggestion.

File: src/common/meta/src/ddl/truncate_table.rs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                    self.data.table_ref(),
+                    region_id,
+                    datanode
+                );
+
+                let request = RegionRequest {
+                    header: Some(RegionRequestHeader {
+                        trace_id: 0,
+                        span_id: 0,
+                    }),
```
Comment (evenyag) [inline src/common/meta/src/ddl/truncate_table.rs:159]: @waynexia  Should we attach the trace id to it like this?
https://github.com/GreptimeTeam/greptimedb/blob/54e506a4940d01ffd3f2e44e450ab6e8cca35481/src/common/meta/src/ddl/alter_table.rs#L207-L209

Only `alter_table` does this. `create_table` and `drop_table` still use 0 as trace id.

Comment (WenyXu) [reply src/common/meta/src/ddl/truncate_table.rs:159]: Should the `trace_id` be retrieved from the Frontend request?

Comment (evenyag) [reply src/common/meta/src/ddl/truncate_table.rs:159]: I think this is a little out-of-topic.

**Review by evenyag — APPROVED** [review]
Rest LGTM.

**Review by WenyXu — APPROVED** [review]

File: src/common/meta/src/ddl/truncate_table.rs (hunk 7 above)
Comment (waynexia) [reply src/common/meta/src/ddl/truncate_table.rs:159]: >Only alter_table does this. create_table and drop_table still use 0 as trace id.

That's not right, can you add it later?

Comment (evenyag) [reply src/common/meta/src/ddl/truncate_table.rs:159]: > > Only alter_table does this. create_table and drop_table still use 0 as trace id.
> 
> That's not right, can you add it later?

@DevilExileSu Could you please add the trace id to the header in this PR?

Comment (DevilExileSu) [reply src/common/meta/src/ddl/truncate_table.rs:159]: Should all headers be added?

**Pull request merged**